internal/app/knitgateway/gateway.go

```diff
@@ -20,49 +20,152 @@ import (
-// (aka "H2C") to send requests to servers.
-var DefaultH2CClient = &http.Client{ //nolint:gochecknoglobals
-	Transport: &http2.Transport{
-		AllowHTTP: true,
-		DialTLSContext: func(ctx context.Context, network, addr string, _ *tls.Config) (net.Conn, error) {
-			return (&net.Dialer{}).DialContext(ctx, network, addr)
+const (
+	clientIdentityHeader = "Knit-Client-Subject"
+)
+
+//nolint:gochecknoglobals
+var (
+	defaultDialer = &net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}
+
+	// defaultH2CClient is like http.DefaultClient except that it will use HTTP/2 over plaintext
+	// (aka "H2C") to send requests to servers.
+	defaultH2CClient = &http.Client{
+		Transport: &http2.Transport{
+			AllowHTTP: true,
+			DialTLSContext: func(ctx context.Context, network, addr string, _ *tls.Config) (net.Conn, error) {
+				return defaultDialer.DialContext(ctx, network, addr)
+			},
 		},
-	},
-}
+	}
+)
 
 // CreateGateway creates a Knit gateway with the given configuration.
 func CreateGateway(config *GatewayConfig) (*knit.Gateway, error) {
 	gateway := &knit.Gateway{}
 	for svcName, svcConf := range config.Services {
-		desc, err := svcConf.Descriptors.FindDescriptorByName(protoreflect.FullName(svcName))
+		httpClient, err := makeHTTPClient(svcConf)
+		if err != nil {
+			return nil, fmt.Errorf("backend #%d: failed to create HTTP client: %w", svcConf.Index+1, err)
+		}
+		descSource := svcConf.Descriptors
+		if deferredSrc, ok := descSource.(*deferredGRPCDescriptorSource); ok {
+			descSource = deferredSrc.WithHTTPClient(httpClient)
+		}
+
+		desc, err := descSource.FindDescriptorByName(protoreflect.FullName(svcName))
 		if err != nil {
 			return nil, fmt.Errorf("could not get descriptor for service %s: %w", svcName, err)
 		}
 		svc, ok := desc.(protoreflect.ServiceDescriptor)
 		if !ok {
 			return nil, fmt.Errorf("%s is a %s, not a service", svcName, descriptorKind(desc))
 		}
-		httpClient := http.DefaultClient
-		if svcConf.H2C {
-			httpClient = DefaultH2CClient
-		}
+		opts := append(svcConf.ConnectOpts, connect.WithInterceptors(&certForwardingInterceptor{})) //nolint:gocritic
 		err = gateway.AddService(
 			svc,
 			knit.WithRoute(svcConf.BaseURL),
 			knit.WithClient(httpClient),
-			knit.WithClientOptions(svcConf.ConnectOpts...),
-			knit.WithTypeResolver(typeResolver{DescriptorSource: svcConf.Descriptors}),
+			knit.WithClientOptions(opts...),
+			knit.WithTypeResolver(typeResolver{DescriptorSource: descSource}),
 		)
 		if err != nil {
 			return nil, fmt.Errorf("failed to register service %s with gateway: %w", svcName, err)
 		}
 	}
 	return gateway, nil
 }
+
+type certForwardingInterceptor struct{}
+
+func (c *certForwardingInterceptor) WrapUnary(unaryFunc connect.UnaryFunc) connect.UnaryFunc {
+	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
+		if cert := clientCertFromContext(ctx); cert != nil {
+			req.Header().Set(clientIdentityHeader, cert.Subject.String())
```

Comment:
It is not immediately apparent to me why this is useful, could you throw some light on why this is needed?

Reply:
If authentication is done via header or cookie, then we already pass that around via header propagation. But authentication is done via TLS cert, then all requests from the gateway look like the same entity: the knit gateway (whatever subject is in its client cert). So we pass along the actual client subject in case downstream services need it for authorization purposes.